File: android/app/src/main/kotlin/ai/gowtam/oak/wire/DexLookup.kt
package ai.gowtam.oak.wire

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Team-builder dex-lookup wire DTOs — read-only enrichment for the entity
 * pickers (species/ability/item/move search, learnset filtering, batch
 * sprites). Faithful mirrors of the three public GET routes:
 *  - `GET /api/search` → `{ matches: SearchMatch[] }`.
 *  - `GET /api/learnset` → `{ moves: LearnsetMove[] }`.
 *  - `GET /api/sprites` → `{ refs: { [name]: DexSpriteRef } }`.
 *
 * All three routes are public (no auth) and never throw for an in-domain miss —
 * an unreadable index degrades to an empty list/map on a 200; the client
 * services (P4) mirror that by folding every transport/decode fault to the same
 * empty result rather than throwing.
 */

/** One typeahead candidate from `GET /api/search` — mirrors `SearchMatch`. */
@Serializable
data class SearchMatch(
    val slug: String,
    @SerialName("display_name") val displayName: String,
    val kind: EntityKind,
)

/**
 * One legal move for a species' movepool from `GET /api/learnset` — mirrors
 * `LearnsetOption`. `type`/`damageClass`/`power` are F1 metadata the moves
 * table renders alongside each picker; they ride along only when the move has
 * cached reference detail, so all three are optional.
 */
@Serializable
data class LearnsetMove(
    val slug: String,
    @SerialName("display_name") val displayName: String,
    val type: String? = null,
    @SerialName("damage_class") val damageClass: DamageClass? = null,
    val power: Int? = null,
) {
    /** The three move damage classes — mirrors `MoveDamageClass`. */
    @Serializable
    enum class DamageClass {
        @SerialName("physical") PHYSICAL,
        @SerialName("special") SPECIAL,
        @SerialName("status") STATUS,
    }
}

/**
 * Batch sprite/type/ability/base-stat ref for one species from
 * `GET /api/sprites` — mirrors `SpriteRef` (`web/src/data/repos/pokedex-repo.ts`).
 * [requiredItem] is the Mega-stone slug the team builder auto-forces onto the
 * held-item field; [abilities] is the form's legal ability slugs (the Ability
 * picker's ONLY offered options). Both are optional so older cached responses
 * without the columns still decode.
 */
@Serializable
data class DexSpriteRef(
    @SerialName("display_name") val displayName: String,
    @SerialName("sprite_url") val spriteUrl: String,
    @SerialName("dex_number") val dexNumber: Int,
    val types: List<String>,
    @SerialName("required_item") val requiredItem: String? = null,
    val abilities: List<String>? = null,
    /** Reuses [BaseStats] — the sprite batch route's `base_stats` is structurally identical. */
    @SerialName("base_stats") val baseStats: BaseStats,
)

File: ios/OakApp/Models/Wire/DexLookup.swift
import Foundation

/// Team-builder dex-lookup wire DTOs — read-only enrichment for the entity pickers
/// (species / ability / item / move search, learnset filtering, batch sprites). Faithful
/// mirrors of the three public GET routes; the TS route handlers are authoritative
/// (CLAUDE.md — "the TS source wins"):
///   - `web/src/app/api/search/route.ts` — `GET /api/search` → `{ matches: SearchMatch[] }`.
///   - `web/src/app/api/learnset/route.ts` — `GET /api/learnset` → `{ moves: LearnsetMove[] }`.
///   - `web/src/app/api/sprites/route.ts` — `GET /api/sprites` → `{ refs: { [name]: DexSpriteRef } }`.
///
/// All three routes are public (no auth — Pokédex reference data) and never throw for an
/// in-domain miss: an unreadable index degrades to an empty list/map on a 200. The three
/// client helpers on `DexLookupService` mirror that by folding every transport/decode
/// fault to the same empty result rather than throwing (`search-client.ts` /
/// `learnset-client.ts` / `sprites-client.ts` never throw either).

/// One typeahead candidate from `GET /api/search` — mirrors `SearchMatch` in
/// `web/src/lib/api/search-client.ts`.
///
/// `sprite_url` is additive and Pokémon-only: present on Pokémon matches when
/// the index has a sprite, omitted (decode → `nil`) when unknown or for
/// moves/abilities/items. Older responses without the key still decode.
struct SearchMatch: Decodable, Sendable, Equatable, Identifiable {
  let slug: String
  let displayName: String
  let kind: EntityKind
  let spriteUrl: String?

  enum CodingKeys: String, CodingKey {
    case slug
    case displayName = "display_name"
    case kind
    case spriteUrl = "sprite_url"
  }

  init(slug: String, displayName: String, kind: EntityKind, spriteUrl: String? = nil) {
    self.slug = slug
    self.displayName = displayName
    self.kind = kind
    self.spriteUrl = spriteUrl
  }

  var id: String { slug }
}

/// One legal move for a species' movepool from `GET /api/learnset` — mirrors
/// `LearnsetOption` in `web/src/lib/api/learnset-client.ts`. `type`/`damageClass`/`power`
/// are the F1 metadata columns the moves table renders alongside each picker; they ride
/// along only when the move has cached reference detail, so all three are optional.
struct LearnsetMove: Decodable, Sendable, Equatable, Identifiable {
  /// The three move damage classes — mirrors `MoveDamageClass`.
  enum DamageClass: String, Decodable, Sendable, Equatable {
    case physical
    case special
    case status
  }

  let slug: String
  let displayName: String
  let type: String?
  let damageClass: DamageClass?
  let power: Int?

  enum CodingKeys: String, CodingKey {
    case slug
    case displayName = "display_name"
    case type
    case damageClass = "damage_class"
    case power
  }

  var id: String { slug }
}

/// Batch sprite/type/ability/base-stat ref for one species from `GET /api/sprites` —
/// mirrors `SpriteRef` in `web/src/data/repos/pokedex-repo.ts`. `requiredItem` is the
/// Mega-stone slug the team builder auto-forces onto the held-item field; `abilities` is
/// the form's legal ability slugs (the Ability picker's ONLY offered options). Both are
/// optional so older cached responses without the columns still decode.
struct DexSpriteRef: Decodable, Sendable, Equatable {
  let displayName: String
  let spriteUrl: String
  let dexNumber: Int
  let types: [String]
  let requiredItem: String?
  let abilities: [String]?
  /// Reuses `BaseStats` (`web/src/agent/schemas.ts` `baseStatsSchema` shape) — the sprite
  /// batch route's `base_stats` field is structurally identical (full stat-name keys).
  let baseStats: BaseStats

  enum CodingKeys: String, CodingKey {
    case displayName = "display_name"
    case spriteUrl = "sprite_url"
    case dexNumber = "dex_number"
    case types
    case requiredItem = "required_item"
    case abilities
    case baseStats = "base_stats"
  }
}
